src/main/scala/dsls/regex/RegularExpression.scala

```diff
@@ -1,16 +1,41 @@
 package dsls.regex
 
-/**
- * Modify this file to implement an internal DSL for regular expressions. 
- * 
- * You're allowed to add anything you want to this file, but you're not allowed
- * to *remove* anything that currently appears in the file.
- */
+import scala.language.implicitConversions
+import scala.language.postfixOps
+
 
 /** The top of a class hierarchy that encodes regular expressions. */
 abstract class RegularExpression {
   /** returns true if the given string matches this regular expression */
   def matches(string: String) = RegexMatcher.matches(string, this)
+
+  /** returns a new RegularExpression representing the union of this
+   *  expression and other
+   */
+  def ||(other: RegularExpression) = Union(this, other)
+
+  /** returns a new RegularExpression representing the concatenation of
+   *  this expression and other
+   */
+  def ~(other: RegularExpression) = Concat(this, other)
+
+  /** returns a new RegularExpression representing the Kleene star of
+   *  this expression, ie EPSILON || this || this~this || ...
+   */
+  def <*> = Star(this)
+
+  /** returns a new RegularExpression which matches one or more
+   *  repetitions of this expression.
+   */
+  def <+> = this ~ (this <*>)
+
+  def apply(n: Int): RegularExpression =
+  if (n == 0)
+    EPSILON
+  else
+    this ~ this{n - 1}
+
+  def apply(regex: RegularExpression): RegularExpression = this ~ regex
```

Comment:
Hmm, what do you need this for? You don't need/want a regex like `p{(o||1<+>)}`.